quickwit/quickwit-proto/src/types/doc_uid.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (C) 2024 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::fmt;
+
+use serde::de::Error;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+pub use ulid::Ulid;
+
+use super::ULID_SIZE;
+
+/// A Doc UID identifies a document across segments, splits, and indexes.
+#[derive(Clone, Copy, Default, Hash, Eq, PartialEq, Ord, PartialOrd)]
+pub struct DocUid(Ulid);
+
+impl fmt::Debug for DocUid {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Doc({})", self.0)
+    }
+}
+
+impl fmt::Display for DocUid {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl From<Ulid> for DocUid {
+    fn from(ulid: Ulid) -> Self {
+        Self(ulid)
+    }
+}
+
+impl DocUid {
+    /// Creates a new random doc UID.
+    pub fn new() -> Self {
+        Self(Ulid::new())
+    }
+
+    #[cfg(any(test, feature = "testsuite"))]
+    pub fn for_test(ulid_u128: u128) -> DocUid {
+        Self(Ulid::from(ulid_u128))
+    }
+
+    /// Returns a new `DocUid` that is greater than the current one.
+    pub fn increment(&self) -> DocUid {
+        // Clippy insists on using `unwrap_or_default`, but that's really not what we want here:
+        // https://github.com/rust-lang/rust-clippy/issues/11631
+        #[allow(clippy::unwrap_or_default)]
+        let next_ulid = self.0.increment().unwrap_or_else(Ulid::new);
+        Self(next_ulid)
+    }
+}
+
+impl<'de> Deserialize<'de> for DocUid {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where D: Deserializer<'de> {
+        let ulid_str: String = String::deserialize(deserializer)?;
```

Comment:
I think the following works.
(I have no clue if we call it a lot or though)

```suggestion
        let ulid_str: Cow<'de, str> = Cow::deserialize(deserializer)?;
```

Reply:
Yes, it works. There are a few other places that could benefit from this pattern.